packages/allow-scripts/src/config.js

```diff
@@ -0,0 +1,358 @@
+    // const canonicalName = getCanonicalNameForPath({ rootDir, filePath: filePath })
+    /** @type {LavamoatPackageJson} */
+    let depPackageJson
+    // eslint-disable-next-line no-useless-catch
+    try {
+      depPackageJson = JSON.parse(
+        await fs.readFile(path.join(filePath, 'package.json'), 'utf-8')
+      )
+    } catch (err) {
+      // FIXME: leftovers of code that used to work
+      // const branchIsOptional = branch.some(node => node.optional)
+      // if (err.code === 'ENOENT' && branchIsOptional) {
+      //   continue
+      // }
+      throw err
+    }
+    const depScripts = depPackageJson.scripts || {}
+    const lifeCycleScripts = ['preinstall', 'install', 'postinstall'].filter(
+      (name) => Object.prototype.hasOwnProperty.call(depScripts, name)
+    )
+
+    if (
+      !lifeCycleScripts.includes('preinstall') &&
+      !lifeCycleScripts.includes('install') &&
+      existsSync(path.join(filePath, 'binding.gyp'))
+    ) {
+      lifeCycleScripts.unshift('install')
+      depScripts.install = 'node-gyp rebuild'
+    }
+
+    if (lifeCycleScripts.length) {
+      /**
+       * @type {{
+       *   canonicalName: string
+       *   path: string
+       *   scripts: LavamoatPackageJson['scripts']
+       * }[]}
+       */
+      const collection = packagesWithScriptsLifecycle.get(canonicalName) || []
+      collection.push({
+        canonicalName,
+        path: filePath,
+        scripts: depScripts,
+      })
+      packagesWithScriptsLifecycle.set(canonicalName, collection)
+    }
+
+    if (FEATURE.bins && depPackageJson.bin) {
+      const binsList = /** @type {[string, string][]} */ (
+        Object.entries(normalizeBin(depPackageJson)?.bin || {})
+      )
+
+      binsList.forEach(([name, link]) => {
+        /** @type {BinInfo[]} */
+        const collection = binCandidates.get(name) || []
+        if (collection.length === 0) {
+          binCandidates.set(name, collection)
+        }
+        collection.push({
+          // canonical name for a direct dependency is just dependency name
+          isDirect: directDeps.has(canonicalName),
+          bin: name,
+          path: filePath,
+          link,
+          fullLinkPath: path.relative(rootDir, path.join(filePath, link)),
+          canonicalName,
+        })
+      })
+    }
+  }
+
+  const lavamoatConfig = packageJson.lavamoat || {}
+
+  const configs = {
+    lifecycle: indexLifecycleConfiguration({
+      packagesWithScripts: packagesWithScriptsLifecycle,
+      allowConfig: lavamoatConfig.allowScripts,
+    }),
+    bin: indexBinsConfiguration({
+      binCandidates,
+      allowConfig: lavamoatConfig.allowBins,
+    }),
+  }
+
+  const somePoliciesAreMissing = !!(
+    configs.lifecycle.missingPolicies.length ||
+    configs.bin.somePoliciesAreMissing
+  )
+
+  return {
+    packageJson,
+    configs,
+    somePoliciesAreMissing,
+    canonicalNamesByPath,
+  }
+}
+
+/**
+ * @typedef GetOptionsForBinOpts
```

Comment:
I consider options to be _optional_, but these are not. Can we call this `GetOptionsForBinParams` instead? I'd like to adopt a naming convention around stuff like this ...

... or avoid it entirely.  FWIW, if I was writing this function, the signature of `getOptionsForBin` would be `(rootDir: string, name: string) => Promise<BinInfo[] | undefined>`, but we're not refactoring here. 😄

Reply:
Fair point and I'd like to do some further cleaning up of some of these typedefs.

This is copied as verbatim, from the existing code being moved, though. Improving on the types outside of facilitating the refactor is something I'd like to do in a follow-up in a separate PR to keep the diff clean.